src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -1413,40 +1423,46 @@ codet java_bytecode_convert_methodt::convert_instructions(
       const bool is_double('d'==type_char);
       const bool is_float('f'==type_char);
 
-      if(is_double || is_float)
+       if(is_double || is_float)
       {
         const ieee_float_spect spec(
           is_float?ieee_float_spect::single_precision():
           ieee_float_spect::double_precision());
 
         ieee_floatt value(spec);
-        const typet &arg_type(arg0.type());
-        if(ID_integer==arg_type.id())
-          value.from_integer(arg0.get_int(ID_value));
+        if(arg0.type().id()!=ID_floatbv)
+        {
+          mp_integer number;
+          if(to_integer(to_constant_expr(arg0), number))
+            throw "failed to convert constant";
+          value.from_integer(number);
+        }
         else
           value.from_expr(to_constant_expr(arg0));
-
         results[0]=value.to_expr();
       }
       else
       {
-        const unsigned value(arg0.get_unsigned_int(ID_value));
+        mp_integer value;
+        if(to_integer(to_constant_expr(arg0), value))
+          throw "failed to convert constant";
         const typet type=java_type_from_char(statement[0]);
         results[0]=from_integer(value, type);
       }
     }
     else if(statement==patternt("?ipush"))
     {
       assert(results.size()==1);
-      mp_integer int_value;
-      bool ret=to_integer(to_constant_expr(arg0), int_value);
-      INVARIANT(!ret, "?ipush argument should be an integer");
-      results[0]=from_integer(int_value, java_int_type());
+      results[0]=arg0;
```

Comment:
This seems to change assumptions on arg0 (need not be a constant nor of java_int_type). Please clarify.